### PR TITLE
Use database as a roomstore

### PR DESCRIPTION
**Pull request by Half-Shot**
This replaces the room store component of the appservice-bridge lib. It uses the same function signatures where possible, but be warned that it only implements what is strictly necessary.

Things to do:
* ~Fix broken tools.~
* ~Fix broken tests.~
* Cache more?
* User store?

(This resolves some of #321 )

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,7 +14,8 @@
     "usertool": "node ./build/tools/userClientTools.js",
     "directoryfix": "node ./build/tools/addRoomsToDirectory.js",
     "ghostfix": "node ./build/tools/ghostfix.js",
-    "chanfix": "node ./build/tools/chanfix.js"
+    "chanfix": "node ./build/tools/chanfix.js",
+    "moveRoomStore": "node ./build/tools/moveRoomStoreToDb.js"
```
Comment (Sorunome) [inline package.json]: Is this tool needed if we have stuff in the migraiton?

File: src/bot.ts [hunk 2]
```diff
@@ -500,9 +500,9 @@ export class DiscordBot {
             throw Error("Room(s) not found.");
         }
         const entry = entries[0];
-        const guild = client.guilds.get(entry.remote.get("discord_guild"));
+        const guild = client.guilds.get(entry.remote!.get("discord_guild") as string);
```
Comment (Sorunome) [inline src/bot.ts]: why are we assuming `entry.remote` is defined? shouldn't we check for this somehow?

Same goes for all the other additions of `!`

File: src/db/connector.ts [hunk 3]
```diff
@@ -14,12 +14,14 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+type SQLTYPES = number | boolean | string | null;
```
Comment (Sorunome) [inline src/db/connector.ts]: Nice, good thinking!

File: src/db/roomstore.ts [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+
+interface IRemoteRoomData extends IRemoteRoomDataLazy {
+    discord_guild: string;
+    discord_channel: string;
+}
+
+interface IRemoteRoomDataLazy  {
+    discord_guild?: string;
+    discord_channel?: string;
+    discord_name?: string|null;
```
Comment (Sorunome) [inline src/db/roomstore.ts]: why do we have both `?` (implying `undefined`) and `null` on these?

File: src/db/roomstore.ts [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+    update_icon?: number|boolean|null;
+    plumbed?: number|boolean|null;
+}
+
+export class RemoteStoreRoom {
+    public data: IRemoteRoomData;
+    constructor(public readonly roomId: string, data: IRemoteRoomData) {
+        for (const k of ["discord_guild", "discord_channel", "discord_name",
+        "discord_topic", "discord_iconurl", "discord_iconurl_mxc", "discord_type"]) {
+            data[k] = typeof(data[k]) === "number" ? String(data[k]) : data[k] || null;
```
Comment (Sorunome) [inline src/db/roomstore.ts]: if `data[k]` doesn't exist, shouldn't we rather set `undefined` than `null`?

File: src/db/roomstore.ts [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+        }
+        for (const k of ["update_name", "update_topic", "update_icon", "plumbed"]) {
+            data[k] = Number(data[k]) || 0;
```
Comment (Sorunome) [inline src/db/roomstore.ts]: These are technically booleans, why do we treat them as numbers?

File: src/db/roomstore.ts [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+
+    public getId() {
+        return this.roomId;
+    }
+}
+
+export interface IRoomStoreEntry {
+    id: string;
+    matrix: MatrixStoreRoom|null;
+    remote: RemoteStoreRoom|null;
```
Comment (Sorunome) [inline src/db/roomstore.ts]: won't some things break if we leave out the shared `data` key, even if we don't use it? Same for `matrix_id` and `remote_id`

File: src/db/roomstore.ts [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+    }
+
+    public async upsertEntry(entry: IRoomStoreEntry) {
+        const promises: Promise<void>[] = [];
+
+        const row = (await this.db.Get("SELECT * FROM room_entries WHERE id = $id", {id: entry.id})) || {};
+
+        if (!row.id) {
+            // Doesn't exist at all, create the room_entries row.
+            const values = {
```
Comment (Sorunome) [inline src/db/roomstore.ts]: `const values: IRoomStoreEntry` (missing type)

File: src/db/roomstore.ts [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+            try {
+                await this.db.Run(`INSERT INTO room_entries VALUES ($id, $matrix, $remote)`, values);
+                log.verbose("Created new entry " + entry.id);
+            } catch (ex) {
+                log.error("Failed to insert room entry", ex);
+                throw Error("Failed to insert room entry");
+            }
+        }
+
+        const matrixId = entry.matrix ? entry.matrix.roomId : null;
```
Comment (Sorunome) [inline src/db/roomstore.ts]: shouldn't this stuff go into an else-block? we unneededly check if we need updating right after we inserted a new entry

File: src/db/roomstore.ts [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+        const remoteId = entry.remote ? entry.remote.roomId : null;
+        const mxIdDifferent = matrixId !== row.matrix_id;
+        const rmIdDifferent = remoteId !== row.remote_id;
+        // Did the room ids change?
+        if (row.id && (mxIdDifferent || rmIdDifferent)) {
```
Comment (Sorunome) [inline src/db/roomstore.ts]: yes, we check again for `row.id` here, but it is kinda....redundant, we do the same check twice

File: src/db/roomstore.ts [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+            "SELECT * FROM room_entries WHERE matrix_id = $id", {id: matrixId},
+        );
+        const res: IRoomStoreEntry[] = [];
+        for (const entry of entries) {
+            const remoteId = entry.remote_id as string || "";
+            const row = await this.db.Get(
+                "SELECT * FROM remote_room_data WHERE room_id = $rid",
+                {rid: remoteId},
+            );
+            if (!row) { continue; }
```
Comment (Sorunome) [inline src/db/roomstore.ts]: why continue? shouldn't we push a thing with blank remote data?

File: src/db/roomstore.ts [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+
+            res.push({
+                id: (entry.id as string),
+                matrix: matrixId ? new MatrixStoreRoom(matrixId) : null,
+                // tslint:disable-next-line no-any
+                remote: remoteId ? new RemoteStoreRoom(remoteId, row as any) : null,
```
Comment (Sorunome) [inline src/db/roomstore.ts]: why `any` and not `IRemoteRoomDataLazy`

File: src/db/roomstore.ts [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+        }
+        if (res.length > 0) {
+            this.entriesMatrixIdCache.set(matrixId, {e: res, ts: Date.now()});
+        }
+        return res;
+    }
+
+    public async getEntriesByMatrixIds(matrixIds: string[]): Promise<IRoomStoreEntry[]> {
+        const entries = await this.db.All(
+            `SELECT * FROM room_entries WHERE matrix_id IN ('${matrixIds.join("','")}')`,
```
Comment (Sorunome) [inline src/db/roomstore.ts]: this doesn't look particularly SQL injection safe. One could easily make a homeserver which creates sql-injection-ready room IDs and poof goes travis' DB

File: src/db/roomstore.ts [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+        );
+        const res: IRoomStoreEntry[] = [];
+        for (const entry of entries) {
+            const matrixId = entry.matrix_id as string || "";
+            const remoteId = entry.remote_id as string || "";
+            const row = await this.db.Get(
+                "SELECT * FROM remote_room_data WHERE room_id = $rid",
+                {rid: remoteId},
+            );
+            if (!row) { continue; }
```
Comment (Sorunome) [inline src/db/roomstore.ts]: same, why not append blank?

File: src/db/roomstore.ts [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+                remote: remoteId ? new RemoteStoreRoom(remoteId, row as any) : null,
+            });
+        }
+        if (res.length > 0) {
+            this.entriesMatrixIdCache.set(matrixId, {e: res, ts: Date.now()});
+        }
+        return res;
+    }
+
+    public async getEntriesByMatrixIds(matrixIds: string[]): Promise<IRoomStoreEntry[]> {
```
Comment (Sorunome) [inline src/db/roomstore.ts]: This function doesn't cache any?

File: src/db/roomstore.ts [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+            log.error("Failed to insert room entry", ex);
+            throw Error("Failed to insert room entry");
+        }
+    }
+
+    public async setMatrixRoom(matrixRoom: MatrixStoreRoom) {
+        // This no-ops, because we don't store anything interesting.
+    }
+
+    public async getEntriesByRemoteRoomData(data: IRemoteRoomDataLazy): Promise<IRoomStoreEntry[]> {
```
Comment (Sorunome) [inline src/db/roomstore.ts]: this whole method feels clunky

File: src/db/roomstore.ts [hunk 17]
```diff
@@ -0,0 +1,347 @@
...
+    }
+
+    public async removeEntriesByRemoteRoomId(remoteId: string) {
+        await this.db.Run(`DELETE FROM room_entries WHERE remote_id = $remoteId`, {remoteId});
+        await this.db.Run(`DELETE FROM remote_room_data WHERE room_id = $remoteId`, {remoteId});
+    }
+
+    public async removeEntriesByMatrixRoomId(matrixId: string) {
+        await this.db.Run(`DELETE FROM room_entries WHERE matrix_id = $matrixId`, {matrixId});
+        await this.db.Run(`DELETE FROM remote_room_data WHERE room_id = $matrixId`, {matrixId});
```
Comment (Sorunome) [inline src/db/roomstore.ts]: eeer this wouldn't find anything? we'd need to select the remote room_id first before deleting it from room_entries?

File: src/db/roomstore.ts [hunk 18]
```diff
@@ -0,0 +1,347 @@
...
+    }
+
+    private async upsertRoom(room: RemoteStoreRoom) {
+        const existingRow = await this.db.Get(
+            "SELECT * FROM remote_room_data WHERE room_id = $id",
+            {id: room.roomId},
+        );
+
+        if (!room.data) {
+            throw new Error("Tried to upsert a room with undefined data");
```
Comment (Sorunome) [inline src/db/roomstore.ts]: we could check for this before performing a db query!

File: src/db/roomstore.ts [hunk 19]
```diff
@@ -0,0 +1,347 @@
...
+
+        const data = {
+            discord_channel:     room.data.discord_channel,
+            discord_guild:       room.data.discord_guild,
+            discord_iconurl:     room.data.discord_iconurl,
+            discord_iconurl_mxc: room.data.discord_iconurl_mxc,
+            discord_name:        room.data.discord_name,
+            discord_topic:       room.data.discord_topic,
+            discord_type:        room.data.discord_type,
+            plumbed:             room.data.plumbed || 0,
```
Comment (Sorunome) [inline src/db/roomstore.ts]: shouldn't these be booleans?

File: src/db/roomstore.ts [hunk 20]
```diff
@@ -0,0 +1,347 @@
...
+            `,
+            {
+                id: room.roomId,
+                // tslint:disable-next-line no-any
+                ...data as any,
+            });
+            return;
+        }
+
+        const keysToUpdate = { };
```
Comment (Sorunome) [inline src/db/roomstore.ts]: this doesn't have a type

File: src/db/schema/v8.ts [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+                await store.roomStore.linkRooms(matrix, remote);
+                log.info(`Migrated ${matrix.roomId}`);
+            } catch (ex) {
+                log.error(`Failed to link ${matrix.roomId}: `, ex);
+            }
+        }
+    }
+
+    public async rollBack(store: DiscordStore): Promise<void> {
+        await store.db.Run(
```
Comment (Sorunome) [inline src/db/schema/v8.ts]: do we want to restore old roomstore on rollback?

File: src/discordas.ts [hunk 22]
```diff
@@ -94,8 +93,17 @@ async function run(port: number, fileConfig: DiscordBridgeConfig) {
                     return await roomhandler.OnAliasQuery.bind(roomhandler)(alias, aliasLocalpart);
                 } catch (err) { log.error("Exception thrown while handling \"onAliasQuery\" event", err); }
             },
-            onEvent: async (request, context) => {
+            onEvent: async (request) => {
                 try {
+                    // Build our own context.
```
Comment (Sorunome) [inline src/discordas.ts]: why do we need this?

File: src/matrixroomhandler.ts [hunk 23]
```diff
@@ -83,14 +86,28 @@ export class MatrixRoomHandler {
...
+    public setBridge(bridge: Bridge, roomStore: DbRoomStore) {
         this.bridge = bridge;
+        this.roomStore = roomStore;
     }
 
     public async OnAliasQueried(alias: string, roomId: string) {
         log.verbose(`Got OnAliasQueried for ${alias} ${roomId}`);
         let channel: Discord.GuildChannel;
         try {
+            // We previously stored the room as an alias.
```
Comment (Sorunome) [inline src/matrixroomhandler.ts]: why not migrate this in the migration?

File: src/provisioner.ts [hunk 24]
```diff
@@ -20,32 +20,32 @@ import {
...
 
 export class Provisioner {
 
-    private bridge: Bridge;
     private pendingRequests: Map<string, (approved: boolean) => void> = new Map(); // [channelId]: resolver fn
+    private roomStore: DbRoomStore;
 
-    public SetBridge(bridge: Bridge): void {
-        this.bridge = bridge;
+    public setStore(roomStore: DbRoomStore) {
```
Comment (Sorunome) [inline src/provisioner.ts]: can't this go in the constructor? with #378 maybe?

File: tools/addRoomsToDirectory.ts [hunk 25]
```diff
@@ -89,20 +90,20 @@ const bridge = new Bridge({
...
     registration: true,
-    roomStore: options.store,
 });
 
+const discordstore = new DiscordStore(config.database ? config.database.filename : "discord.db");
+
 async function run() {
     try {
-        await bridge.loadDatabases();
+        await discordstore.init();
```
Comment (Sorunome) [inline tools/addRoomsToDirectory.ts]: with #378 this should go into that discorbot.init

File: package.json (hunk 1 above)
Comment (Half-Shot) [reply package.json]: No, and I deleted it. I should also remove it from package.json

File: src/db/roomstore.ts (hunk 4 above)
Comment (Half-Shot) [reply src/db/roomstore.ts]: It's a confusing mess :/. We need these to be undefined so that when you want to search for a val, you can leave most of these undefined. However, you may also want to set it to null to explicitly look for null values (since SQLITE uses null to determine an empty value).

Tl;dr - undefined == don't search for this. null == look for a empty value.

File: src/db/roomstore.ts (hunk 5 above)
Comment (Half-Shot) [reply src/db/roomstore.ts]: Nope, because undefined doesn't exist in SQL terms but null means empty.

File: src/db/roomstore.ts (hunk 6 above)
Comment (Half-Shot) [reply src/db/roomstore.ts]: SQLITE has no bool type, so we have to do conversions :/

File: src/db/roomstore.ts (hunk 7 above)
Comment (Half-Shot) [reply src/db/roomstore.ts]: Indeed, this PR is doing the bare minimum to make everything work. I'd expect us to implement matrix and link data when it becomes required. Obviously typescript won't let us just set values without any typing information, anyway.

File: src/db/roomstore.ts (hunk 9 above)
Comment (Half-Shot) [reply src/db/roomstore.ts]: Maybe, but I think this was a tradeoff between a few nano seconds and less indent/less ugly code.

File: src/db/roomstore.ts (hunk 10 above)
Comment (Half-Shot) [reply src/db/roomstore.ts]: oh derp

File: src/db/roomstore.ts (hunk 11 above)
Comment (Half-Shot) [reply src/db/roomstore.ts]: oh um...probably

File: src/db/roomstore.ts (hunk 12 above)
Comment (Half-Shot) [reply src/db/roomstore.ts]: I suspect I forgot about it.

File: src/db/roomstore.ts (hunk 13 above)
Comment (Half-Shot) [reply src/db/roomstore.ts]: Oh balls, yes you are right (though I hoped that sqlite3 / pg might be doing minimal escaping here).

We should validate that the matrixId really is an id (although I use aliases briefly in MRH).

File: src/db/roomstore.ts (hunk 15 above)
Comment (Half-Shot) [reply src/db/roomstore.ts]: I don't think we use it in a high traffic area, so I avoided over optimizing it.

File: src/db/roomstore.ts (hunk 16 above)
Comment (Half-Shot) [reply src/db/roomstore.ts]: I don't think it's that bad? We build up a WHERE clause based on provided data and iterate over the result. Suggestions welcome though, since I might be missing something.

File: src/db/roomstore.ts (hunk 17 above)
Comment (Half-Shot) [reply src/db/roomstore.ts]: GAH, yes. Probably sensible to do a SELECT, then two deletes.

File: src/db/roomstore.ts (hunk 19 above)
Comment (Half-Shot) [reply src/db/roomstore.ts]: Same as aboove.

File: src/db/roomstore.ts (hunk 20 above)
Comment (Half-Shot) [reply src/db/roomstore.ts]: I suspect I was being lazy, this should probably be `IWhatItsNameLazy`

File: src/db/schema/v8.ts (hunk 21 above)
Comment (Half-Shot) [reply src/db/schema/v8.ts]: Well, we don't delete any values from the roomstore. The bridge won't function without it though, and trying to write code to if else around the room stores is going to be a nightmare.

File: src/db/roomstore.ts (hunk 13 above)
Comment (Sorunome) [reply src/db/roomstore.ts]: Or we could create an array of parameters `WHERE matrix_id IN ($1, $2, $3, ....)` and then pass those on

File: src/discordas.ts (hunk 22 above)
Comment (Half-Shot) [reply src/discordas.ts]: We used to use the nedb context in a few places to find out which remote room we were handling, now we do the same thing with the db.

File: src/matrixroomhandler.ts (hunk 23 above)
Comment (Half-Shot) [reply src/matrixroomhandler.ts]: huh? No, we have to store the matrix_id as an alias in `createMatrixRoom` because we don't know the roomId yet.

File: src/db/roomstore.ts (hunk 16 above)
Comment (Sorunome) [reply src/db/roomstore.ts]: not really sure, it just... *feels* clunkly and hard to cache if something happens down the road. Probably fine

File: src/provisioner.ts (hunk 24 above)
Comment (Half-Shot) [reply src/provisioner.ts]: Probably with #378 :)

File: tools/addRoomsToDirectory.ts (hunk 25 above)
Comment (Half-Shot) [reply tools/addRoomsToDirectory.ts]: Ditto!

File: src/db/roomstore.ts (hunk 8 above)
Comment (Half-Shot) [reply src/db/roomstore.ts]: It's not, as it turns out because we are inserting this into the DB which wants strings.

File: src/db/roomstore.ts [hunk 26]
```diff
@@ -0,0 +1,365 @@
+/*
+Copyright 2018 matrix-appservice-discord
```
Comment (turt2live) [inline src/db/roomstore.ts]:
```suggestion
Copyright 2019 matrix-appservice-discord
```

File: src/db/schema/v8.ts [hunk 27]
```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2018 matrix-appservice-discord
```
Comment (turt2live) [inline src/db/schema/v8.ts]:
```suggestion
Copyright 2019 matrix-appservice-discord
```

File: test/db/test_roomstore.ts [hunk 28]
```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2017, 2018 matrix-appservice-discord
```
Comment (turt2live) [inline test/db/test_roomstore.ts]:
```suggestion
Copyright 2019 matrix-appservice-discord
```

File: src/db/roomstore.ts [hunk 29]
```diff
@@ -0,0 +1,365 @@
...
+        const res: IRoomStoreEntry[] = [];
+        for (const entry of entries) {
+            let remote: RemoteStoreRoom|null = null;
+            if (entry.remote_id) {
+                const remoteId = entry.remote_id as string;
+                const row = await this.db.Get(
+                    "SELECT * FROM remote_room_data WHERE room_id = $remoteId",
+                    {remoteId},
+                );
+                if (row) {
```
Comment (Sorunome) [inline src/db/roomstore.ts]: actually, thinking about it more it might be a good idea to remove the entries without a remote. That should kill #225 once and for all as it *should* always be present

File: src/db/roomstore.ts [hunk 30]
```diff
@@ -0,0 +1,365 @@
...
+        for (const entry of entries) {
+            let remote: RemoteStoreRoom|null = null;
+            const matrixId = entry.matrix_id as string || "";
+            const remoteId = entry.remote_id as string;
+            if (remoteId) {
+                const row = await this.db.Get(
+                    "SELECT * FROM remote_room_data WHERE room_id = $rid",
+                    {rid: remoteId},
+                );
+                if (row) {
```
Comment (Sorunome) [inline src/db/roomstore.ts]: same here, maybe it is actually a good idea to ignore results without remote data (like you did originally)

File: tools/addRoomsToDirectory.ts (hunk 25 above)
Comment (Sorunome) [reply tools/addRoomsToDirectory.ts]: also this still
